File: decompiled/c/450b4c.c

void FUN_00450b4c(void)

{
  byte bVar1;
  short sVar2;
  byte bVar3;
  short extraout_DX;
  short sVar4;
  ushort uVar5;
  int iVar6;
  int iVar7;
  uint uVar8;
  int iVar9;
  uint uVar10;
  int iVar11;
  
  bVar3 = 0;
  do {
    uVar10 = (uint)bVar3;
    iVar11 = uVar10 * 0x4b0c;
    bVar1 = (&DAT_008ae9c4)[iVar11];
    if ((bVar1 != 0xff) &&
       (iVar7 = (uint)bVar1 * 0x260, ((&DAT_00887422)[(uint)bVar1 * 0x130] & 1) != 0)) {
      if (((&DAT_008ae9c5)[iVar11] & 1) == 0) {
        for (iVar6 = 0; (byte)iVar6 < (byte)(&DAT_00887498)[iVar7]; iVar6 = iVar6 + 1) {
          if ((*(ushort *)(&DAT_0088747e + iVar6 * 2 + iVar7) != 0xffff) &&
             (iVar9 = (uint)*(ushort *)(&DAT_0088747e + iVar6 * 2 + iVar7) * 0x100,
             (&DAT_00743be4)[iVar9] == '\x03')) {
            (&DAT_008ae9ce)[iVar11] = (byte)iVar6;
            (&DAT_008ae9cf)[iVar11] = (&DAT_00743bdf)[iVar9];
            (&DAT_008ae9c5)[iVar11] = (&DAT_008ae9c5)[iVar11] | 1;
            (&DAT_008ae9c5)[iVar11] = (&DAT_008ae9c5)[iVar11] & 0xfd;
            goto LAB_00450bd4;
          }
        }
      }
      else {
LAB_00450bd4:
        uVar5 = *(ushort *)(&DAT_0088747e + (uint)(byte)(&DAT_008ae9ce)[iVar11] * 2 + iVar7);
        if (uVar5 != 0xffff) {
          iVar6 = (uint)uVar5 * 0x100;
          uVar8 = (uint)(ushort)(&DAT_008ae9cc)[uVar10 * 0x2586];
          if (((&DAT_008ae9c5)[iVar11] & 2) != 0) {
            if ((&DAT_00743be4)[iVar6] != '\x03') goto LAB_00450d74;
            (&DAT_008ae9c5)[iVar11] = (&DAT_008ae9c5)[iVar11] & 0xfd;
            if ((&DAT_00743bdf)[iVar6] == (&DAT_008ae9cf)[iVar11]) {
              (&DAT_008ae9cc)[uVar10 * 0x2586] = 0;
              uVar8 = 0;
            }
          }
          if ((&DAT_00743be4)[iVar6] == '\x06') {
            (&DAT_008ae9c5)[iVar11] = (&DAT_008ae9c5)[iVar11] | 2;
          }
          else {
            if (uVar8 < 0x12c0) {
              if (((&DAT_008ae9c5)[iVar11] & 4) != 0) {
                sVar2 = FUN_005d9220(uVar8,iVar7);
                sVar2 = sVar2 >> 3;
                sVar4 = extraout_DX >> 3;
                if (sVar2 < -0x7e) {
                  sVar2 = -0x7f;
                }
                if (0x7e < sVar2) {
                  sVar2 = 0x7f;
                }
                if (sVar4 < -0x7e) {
                  sVar4 = -0x7f;
                }
                if (0x7e < sVar4) {
                  sVar4 = 0x7f;
                }
                if ((DAT_006e3b84 & 1) == 0) {
                  (&DAT_008ae9d0)[iVar11 + uVar8] = (char)sVar2;
                  (&DAT_008afc90)[iVar11 + uVar8] = (char)sVar4;
                }
                else {
                  (&DAT_008ae9d0)[iVar11 + uVar8] =
                       (char)((short)(sVar2 + (char)(&DAT_008ae9d0)[iVar11 + uVar8]) >> 1);
                  (&DAT_008afc90)[iVar11 + uVar8] =
                       (char)((short)(sVar4 + (char)(&DAT_008afc90)[iVar11 + uVar8]) >> 1);
                }
              }
              iVar7 = *(int *)(&DAT_00743bbc + iVar6) * 5 >> 0x10;
              if (iVar7 < 0) {
                iVar7 = -iVar7;
              }
              uVar5 = (short)(&DAT_00743ba6)[(uint)uVar5 * 0x80] >> 2;
              if (0xfe < uVar5) {
                uVar5 = 0xff;
              }
              if ((DAT_006e3b84 & 1) == 0) {
                (&DAT_008b0f50)[iVar11 + uVar8] = (char)iVar7;
                (&DAT_008b2210)[iVar11 + uVar8] = (char)uVar5;
              }
              else {
                (&DAT_008b0f50)[iVar11 + uVar8] =
                     (char)((ushort)((short)iVar7 + (ushort)(byte)(&DAT_008b0f50)[iVar11 + uVar8])
                           >> 1);
                (&DAT_008b2210)[iVar11 + uVar8] =
                     (char)((ushort)(uVar5 + (byte)(&DAT_008b2210)[iVar11 + uVar8]) >> 1);
              }
              if (((DAT_006e3b84 & 1) != 0) &&
                 (uVar5 = (short)uVar8 + 1, (ushort)(&DAT_008ae9ca)[uVar10 * 0x2586] < uVar5)) {
                (&DAT_008ae9ca)[uVar10 * 0x2586] = uVar5;
              }
            }
            if ((DAT_006e3b84 & 1) != 0) {
              (&DAT_008ae9cc)[uVar10 * 0x2586] = (&DAT_008ae9cc)[uVar10 * 0x2586] + 1;
            }
          }
        }
      }
    }
LAB_00450d74:
    bVar3 = bVar3 + 1;
    if (7 < bVar3) {
      return;
    }
  } while( true );
}

File: ported/auto/450b4c.js
// Auto-translated from Ghidra C by tools/c-to-js/translate.js.
// Source: decompiled/c/450b4c.c
// Edit by hand only after diff-test passes — re-running the translator will overwrite.

/** @typedef {import("../../runtime/heap.js").Heap} Heap */

import { regs } from "../../runtime/regs.js";
import { FUN_005d9220 } from "./5d9220.js";
export function FUN_00450b4c(heap) {
  let bVar1 = 0;
  let sVar2 = 0;
  let bVar3 = 0;
  let extraout_DX = 0;
  let sVar4 = 0;
  let uVar5 = 0;
  let iVar6 = 0;
  let iVar7 = 0;
  let uVar8 = 0;
  let iVar9 = 0;
  let uVar10 = 0;
  let iVar11 = 0;
  bVar3 = ((0) & 0xff);
  do {
    LAB_00450d74: {
    uVar10 = ((((bVar3) >>> 0)) >>> 0);
    iVar11 = ((uVar10 * 0x4b0c) >>> 0);
    bVar1 = ((heap.u32((0x008ae9c4) + (iVar11) * 4)) & 0xff);
    if ((bVar1 != 0xff) && (iVar7 = ((((bVar1) >>> 0) * 0x260) >>> 0), (heap.u32((0x00887422) + (((bVar1) >>> 0) * 0x130) * 4) & 1) != 0)) {
      if ((heap.u32((0x008ae9c5) + (iVar11) * 4) & 1) == 0) {
        for (iVar6 = ((0) >>> 0); ((iVar6) & 0xff) < heap.u32(((0x00887498) & 0xff) + (iVar7) * 4); iVar6 = (((iVar6 + 1) >>> 0)) >>> 0) {
          if ((heap.u16((0x0088747e + iVar6 * 2 + iVar7)) != 0xffff) && (iVar9 = ((heap.u32((0x0088747e + iVar6 * 2 + iVar7)) * 0x100) >>> 0), heap.u32((0x00743be4) + (iVar9) * 4) == 3)) {
            heap.setU32(((0x008ae9ce) + (iVar11) * 4), (((iVar6) & 0xff)) & 0xffffffff);
            heap.setU32(((0x008ae9cf) + (iVar11) * 4), (heap.u32((0x00743bdf) + (iVar9) * 4)) & 0xffffffff);
            heap.setU32(((0x008ae9c5) + (iVar11) * 4), (heap.u32((0x008ae9c5) + (iVar11) * 4) | 1) & 0xffffffff);
            heap.setU32(((0x008ae9c5) + (iVar11) * 4), (heap.u32((0x008ae9c5) + (iVar11) * 4) & 0xfd) & 0xffffffff);
            /* goto LAB_00450bd4 — unsupported, early-return */ if (typeof globalThis._gotoWarn !== 'undefined') globalThis._gotoWarn("FUN_00450b4c/LAB_00450bd4"); return 0;
          }
        }
      } else {
        LAB_00450bd4: uVar5 = ((heap.u16((0x0088747e + heap.u32(((0x008ae9ce) >>> 0) + (iVar11) * 4) * 2 + iVar7))) & 0xffff);
        if (uVar5 != 0xffff) {
          iVar6 = ((((uVar5) >>> 0) * 0x100) >>> 0);
          uVar8 = ((heap.u32(((0x008ae9cc) >>> 0) + (uVar10 * 0x2586) * 4)) >>> 0);
          if ((heap.u32((0x008ae9c5) + (iVar11) * 4) & 2) != 0) {
            if (heap.u32((0x00743be4) + (iVar6) * 4) != 3) {
              break LAB_00450d74;
            }
            heap.setU32(((0x008ae9c5) + (iVar11) * 4), (heap.u32((0x008ae9c5) + (iVar11) * 4) & 0xfd) & 0xffffffff);
            if (heap.u32((0x00743bdf) + (iVar6) * 4) == heap.u32((0x008ae9cf) + (iVar11) * 4)) {
              heap.setU32(((0x008ae9cc) + (uVar10 * 0x2586) * 4), (0) & 0xffffffff);
              uVar8 = ((0) >>> 0);
            }
          }
          if (heap.u32((0x00743be4) + (iVar6) * 4) == 6) {
            heap.setU32(((0x008ae9c5) + (iVar11) * 4), (heap.u32((0x008ae9c5) + (iVar11) * 4) | 2) & 0xffffffff);
          } else {
            if (uVar8 < 0x12c0) {
              if ((heap.u32((0x008ae9c5) + (iVar11) * 4) & 4) != 0) {
                sVar2 = (((regs.eax = FUN_005d9220(heap, uVar8, iVar7))) & 0xffff);
                sVar2 = ((sVar2 >>> 3) & 0xffff);
                sVar4 = ((extraout_DX >>> 3) & 0xffff);
                if (sVar2 < -0x7e) {
                  sVar2 = ((-0x7f) & 0xffff);
                }
                if (0x7e < sVar2) {
                  sVar2 = ((0x7f) & 0xffff);
                }
                if (sVar4 < -0x7e) {
                  sVar4 = ((-0x7f) & 0xffff);
                }
                if (0x7e < sVar4) {
                  sVar4 = ((0x7f) & 0xffff);
                }
                if ((heap.u32(0x006e3b84) & 1) == 0) {
                  heap.setU32(((0x008ae9d0) + (iVar11 + uVar8) * 4), (((sVar2) << 24 >> 24)) & 0xffffffff);
                  heap.setU32(((0x008afc90) + (iVar11 + uVar8) * 4), (((sVar4) << 24 >> 24)) & 0xffffffff);
                } else {
                  heap.setU32(((0x008ae9d0) + (iVar11 + uVar8) * 4), (((((((sVar2 + ((heap.u32((0x008ae9d0) + (iVar11 + uVar8) * 4)) << 24 >> 24))) << 16 >> 16) >>> 1)) << 24 >> 24)) & 0xffffffff);
                  heap.setU32(((0x008afc90) + (iVar11 + uVar8) * 4), (((((((sVar4 + ((heap.u32((0x008afc90) + (iVar11 + uVar8) * 4)) << 24 >> 24))) << 16 >> 16) >>> 1)) << 24 >> 24)) & 0xffffffff);
                }
              }
              iVar7 = ((heap.i32((0x00743bbc + iVar6)) * 5 >>> 0x10) >>> 0);
              if (iVar7 < 0) {
                iVar7 = ((-iVar7) >>> 0);
              }
              uVar5 = ((((heap.u32((0x00743ba6) + (((uVar5) >>> 0) * 0x80) * 4)) << 16 >> 16) >>> 2) & 0xffff);
              if (0xfe < uVar5) {
                uVar5 = ((0xff) & 0xffff);
              }
              if ((heap.u32(0x006e3b84) & 1) == 0) {
                heap.setU32(((0x008b0f50) + (iVar11 + uVar8) * 4), (((iVar7) << 24 >> 24)) & 0xffffffff);
                heap.setU32(((0x008b2210) + (iVar11 + uVar8) * 4), (((uVar5) << 24 >> 24)) & 0xffffffff);
              } else {
                heap.setU32(((0x008b0f50) + (iVar11 + uVar8) * 4), ((((((((iVar7) << 16 >> 16) + heap.u32(((0x008b0f50) & 0xffff) + (iVar11 + uVar8) * 4)) & 0xffff) >>> 1)) << 24 >> 24)) & 0xffffffff);
                heap.setU32(((0x008b2210) + (iVar11 + uVar8) * 4), ((((((uVar5 + heap.u32(((0x008b2210) & 0xff) + (iVar11 + uVar8) * 4)) & 0xffff) >>> 1)) << 24 >> 24)) & 0xffffffff);
              }
              if (((heap.u32(0x006e3b84) & 1) != 0) && (uVar5 = ((((uVar8) << 16 >> 16) + 1) & 0xffff), heap.u32(((0x008ae9ca) & 0xffff) + (uVar10 * 0x2586) * 4) < uVar5)) {
                heap.setU32(((0x008ae9ca) + (uVar10 * 0x2586) * 4), (uVar5) & 0xffffffff);
              }
            }
            if ((heap.u32(0x006e3b84) & 1) != 0) {
              heap.setU32(((0x008ae9cc) + (uVar10 * 0x2586) * 4), (heap.u32((0x008ae9cc) + (uVar10 * 0x2586) * 4) + 1) & 0xffffffff);
            }
          }
        }
      }
    }
    }
    bVar3 = ((bVar3 + 1) & 0xff);
    if (7 < bVar3) {
      return;
    }
  } while (true);
}
